### PR TITLE
[Refactor] Masternode Budget first refactoring and cleanup

### PR DESCRIPTION
**Pull request by random-zebra**
Refactoring of the masternode budget classes (`CBudgetManager`, `CBudgetProposal`, `CBudgetVote`, `CFinalizedBudget`, `CFinalizedBudgetVote`), providing better encapsulation of the data, removing unused functions, and giving a first round of styling cleanup.

### CODE REVIEW

File: src/masternode-budget.cpp [hunk 1]
```diff
@@ -1475,10 +1467,12 @@ bool CBudgetProposal::IsValid(std::string& strError, bool fCheckCollateral)
 
     // if (GetBlockEnd() < pindexPrev->nHeight - GetBudgetPaymentCycleBlocks() / 2) {
     if(nProposalEnd < pindexPrev->nHeight){
-        strError = "Proposal " + strProposalName + ": Invalid nBlockEnd (" + std::to_string(nProposalEnd) + ") < current height (" + std::to_string(pindexPrev->nHeight) + ")";
+        strInvalid = "Proposal " + strProposalName + ": Invalid nBlockEnd (" + std::to_string(nProposalEnd) + ") < current height (" + std::to_string(pindexPrev->nHeight) + ")";
         return false;
     }
 
+    fValid = true;
+    strInvalid = "";
```
Comment (furszy) [inline src/masternode-budget.cpp:1494]: minor nit: `strInvalid.clear()`

File: src/masternode-budget.cpp [hunk 2]
```diff
@@ -2069,10 +2066,12 @@ bool CFinalizedBudget::IsValid(std::string& strError, bool fCheckCollateral)
     int nMaxAge = nBlockStart - (2 * nBlocksPerCycle);
 
     if (GetBlockEnd() < nMaxAge) {
-        strError = strprintf("Budget " + strBudgetName + " (" + strProposals + ") (ends at block %ld) too old and obsolete", GetBlockEnd());
+        strInvalid = strprintf("Budget " + strBudgetName + " (" + strProposals + ") (ends at block %ld) too old and obsolete", GetBlockEnd());
         return false;
     }
 
+    fValid = true;
+    strInvalid = "";
```
Comment (furszy) [inline src/masternode-budget.cpp:2107]: minor nit: `strInvalid.clear()`

File: src/masternode-budget.cpp [hunk 3]
```diff
@@ -457,18 +457,8 @@ bool CBudgetManager::AddProposal(CBudgetProposal& budgetProposal)
...
-        TRY_LOCK(cs_main, locked);
-        if ((locked) && (chainActive.Tip() != NULL)) {
-            CBlockIndex* pindexPrev = chainActive.Tip();
-            if (pindexPrev) {
-                nHeight = pindexPrev->nHeight;
-            }
-        }
-    }
+    int nHeight = WITH_LOCK(cs_main, return chainActive.Height());
+    if (nHeight <= 0) return;
```
Comment (furszy) [inline src/masternode-budget.cpp:461]: let's better remove these two lines.
The check is not needed for the internal procedure, should be checked on the caller's side if needed and most importantly, can be a deadlocking cause.

The callers are:

 `CBudgetManager::NewBlock` --> which doesn't need it. Method only called when a new block arrives. Height will be never 0.

`CBudgetDB::Read` --> only called during init. A check flag can be set before the `LOCK(objToLoad.cs);` line to call or not `CheckAndRemove` below.

`checkbudgets` --> RPC command. Can be checked before calling the method.

File: src/masternode-budget.cpp [hunk 4]
```diff
@@ -1808,6 +1813,17 @@ std::string CBudgetVote::GetStrMessage() const
             std::to_string(nVote) + std::to_string(nTime);
 }
 
+UniValue CBudgetVote::ToJSON() const
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/masternode-budget.cpp:1744]: Yes, or in rpc/budget, where we have, for example, also 
```cpp
void budgetToJSON(CBudgetProposal* pbudgetProposal, UniValue& bObj)
```

But I think it's better to have them as class methods (same as `ToString()`) to be able access freely all private members.

File: src/masternode-budget.h [hunk 5]
```diff
@@ -85,6 +88,18 @@ class CBudgetVote : public CSignedMessage
     std::string GetStrMessage() const override;
     const CTxIn GetVin() const override { return vin; };
 
+    UniValue ToJSON() const;
+
+    int GetDirection() const { return nVote; }
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/masternode-budget.h]: I'm on the fence. Didn't like this "direction" name either... but I think that neither `GetVote` nor `GetVoteType` truly catch the meaning. Open to hear more opinions.

File: src/masternode-budget.cpp (hunk 3 above)
Comment (random-zebra) [reply src/masternode-budget.cpp]: Good Idea.
For the DB read, we can just modify the `fDryRun` flag in init. For the RPC function, I actually think it's better to check before for complete mnsync anyway.
Updated this in e09f3164d78b52091f9158e6d14322cb48c79544

File: src/masternode-budget.cpp (hunk 1 above)
Comment (random-zebra) [reply src/masternode-budget.cpp]: Done

File: src/masternode-budget.cpp (hunk 2 above)
Comment (random-zebra) [reply src/masternode-budget.cpp]: Done

**Comment by random-zebra** [comment]
Addressed @furszy's comments, and added a couple more commits that completely remove any direct `chainActive` access from budget classes.

**Review by furszy — COMMENTED** [review]
Strong concept ACK,

Have tested it with #1829 on top and all good 🕺 

Screen Shot 2020-09-04 at 7 57 24 PM

**Review by furszy — APPROVED** [review]
ACK 08d65f0 and tested it with 1829 latest changes on top and all good.

-- merge order wise: this one need to be merged after #1835 --

**Comment by random-zebra** [comment]
Rebased on master (#1828 and #1835 merged).

File: src/masternode-budget.cpp [hunk 6]
```diff
@@ -951,24 +832,87 @@ CAmount CBudgetManager::GetTotalBudget(int nHeight)
...
+{
+    SetBestHeight(height);
 
     if (masternodeSync.RequestedMasternodeAssets <= MASTERNODE_SYNC_BUDGET) return;
 
     if (strBudgetMode == "suggest") { //suggest the budget we see
         SubmitFinalBudget();
     }
 
+    int nCurrentHeight = GetBestHeight();
```
Comment (furszy) [inline src/masternode-budget.cpp:908]: nit: `int nCurrentHeight = height;

**Review by furszy — APPROVED** [review]
ACK 5564002 and merging..

**Pull request merged**